### PR TITLE
Update feature_extractor.py

### PR DESCRIPTION
**Pull request by BBC-Esq**
Added mel filter bank caching to FeatureExtractor class to optimize memory usage and reduce computational overhead when processing multiple audio files with identical parameters, particularly beneficial for batch processing scenarios.

### CODE REVIEW

File: faster_whisper/feature_extractor.py [hunk 1]
```diff
@@ -54,61 +77,88 @@ def get_mel_filters(sr, n_fft, n_mels=128):
...
-        )
+        # Move waveform to the target device if necessary
+        if self.device == "cuda" and not waveform.is_cuda:
+            waveform = waveform.to(self.device)
 
+        # Apply padding if required
         if padding:
-            waveform = torch.nn.functional.pad(waveform, (0, self.n_samples))
-
-        window = torch.hann_window(self.n_fft).to(waveform.device)
```
Comment (abodacs) [inline faster_whisper/feature_extractor.py:99]: Why was the `hann_window` deleted?

File: faster_whisper/feature_extractor.py [hunk 2]
```diff
@@ -54,61 +77,88 @@ def get_mel_filters(sr, n_fft, n_mels=128):
...
         if waveform.dtype is not torch.float32:
             waveform = waveform.to(torch.float32)
 
-        waveform = (
-            waveform.to(self.device)
-            if self.device == "cuda" and not waveform.is_cuda
-            else waveform
-        )
+        # Move waveform to the target device if necessary
+        if self.device == "cuda" and not waveform.is_cuda:
```
Comment (abodacs) [inline faster_whisper/feature_extractor.py:136]: Improved readability, thank you.

**Review by abodacs — CHANGES_REQUESTED** [review]
@BBC-Esq 
Thanks for your work

File: faster_whisper/feature_extractor.py (hunk 1 above)
Comment (BBC-Esq) [reply faster_whisper/feature_extractor.py:99]: I'll take a look...

Comment (BBC-Esq) [reply faster_whisper/feature_extractor.py:99]: This is actually an optimization. In the old version, a new Hann window was being created and moved to the device every time __call__ was executed. The new version creates it once during initialization and caches it as an instance variable (self.window).

**Comment by MahmoudAshraf97** [comment]
Hello, the way you implemented caching is not going to work because the only parameter that might change when creating is `n_mels`, to change this parameter you need to initialize a new `FeatureExtractor` instance which will clear cache thus invalidating its purpose. Anyways, creating window takes around 6us which represents less than 0.05% of the execution time and the mel filters is created only once thus there is no need for caching.

In the feature, it's preferable to use a caching decorator such as `functools.lru_cache` as an easy caching solution instead of implementing it as a dictionary

**Pull request closed without merging**